specs/eventing/data-plane.md

```diff
@@ -1,125 +1,171 @@
-event replies will be accepted. While a source sends events without the header,
-in which case the sink may assume that any event reply will be dropped without
-error or retry attempt. If a sink wishes to ensure the reply events will be
-delivered, it can check for the existence of the `Prefer: reply` header in the
-delivery request and respond with an error code if the header is not present.
-
-### Data plane contract for Sources
-
-See [Source Delivery specification](sources.md#source-event-delivery)
-for details.
-
-### Data plane contract for Channels
-
-See [Channel Delivery specification](channel.md#data-plane) for details.
-
-### Data plane contract for Brokers
-
-See [Broker Delivery specification](broker.md)
-
-## Changelog
+Event senders MAY add or update CloudEvents attributes before sending to
+implement observability features such as tracing; in particular, the
+[`traceparent` and `tracestate` distributed tracing
+attributes](https://github.com/cloudevents/spec/blob/v1.0/extensions/distributed-tracing.md)
+may be modified in this way for each delivery attempt of the same event.
+
+This specification does not mandate any particular logging or metrics
+aggregtion, nor a method of exposing observability information to users
+configuring the resources. Platform administrators SHOULD expose event-delivery
+telemetry to users through platform-specific interfaces, but such interfaces are
+beyond the scope of this document.
+
+<!-- TODO: should we mention RECOMMENDED spans or RECOMMENDED metrics like in
+https://github.com/knative/specs/blob/main/specs/eventing/channel.md#observability?
+-->
+
+### Derived (Reply) Events
+
+In some applications, an event receiver might emit an event in reaction to a
+received event. An event sender MAY document support for this pattern by
+including a `Prefer: reply` header in the HTTP POST request. This header
+indicates to the event receiver that the caller will accept a [`200`
+response](#event-acknowledgement-and-repeat-delivery) which includes a
+CloudEvent encoded using the binary or structured formats.
+[Brokers](./overview.md#broker) and [Channels](./overview.md#channel) MUST
+indicate support for replies using the `Prefer: reply` header.
+
+The sender SHOULD NOT assume that a received reply event is directly related to
+the event sent in the HTTP request.
+
+A recipient MAY reply to any HTTP POST with a `200` response to indicate that
+the event was processed successfully, with or without a response payload. If the
+recipient will _never_ provide a response payload, the `202` response code is
+likely a better choice.
+
+If a recipient chooses to reply to a sender with a `200` response code and a
+reply event in the absence of a `Prefer: reply` header, the sender SHOULD treat
+the event as accepted, and MAY log an error about the unexpected payload. The
+sender MUST NOT process the reply event if it did not advertise the `Prefer:
```

Comment:
this section brings the `Prefer: reply` to a mandatory and I thought before it was more of a "you can provide this, but the sink is free to ignore it, but if you say "Prefer: " and get a reply, you are free to drop the event without warning. 

This language makes the header mandatory, and it is not used anywhere yet, so it would be an API change for eventing at the moment.

Reply:
Yes, this attempts to make the expected behavior for e.g. a receiving function explicit.

I'm willing to change this, but the current formulation implemented in eventing makes it difficult for a function to know if the reply might be dropped, which seems to go against the goal of avoiding message loss.

Reply:
(I didn't change this yet, because I'd like to discuss whether it makes more sense to change the contract or the implementation.)

Reply:
FYI, here's some more background on this.
https://github.com/knative/eventing/pull/4560

Reply:
I see the background, and the concern that adding `MUST` language around this header would make previous versions of Knative incompatible with the 1.0 spec.

https://github.com/knative/eventing/pull/4560#issuecomment-759774906

I'm wondering if we have a requirement that versions of Knative prior to the 1.0 spec `MUST` be compliant? If so, what is our boundary?

Reply:
Given that this functionality is not written in the reference implementation, I would be inclined to not make it a `MUST`.

The original PR was using `SHOULD` and `MAY`, no `MUST NOT` nor `MUST`

Reply:
The original PR did not include any MUST clauses.

Unfortunately, I have a conflict with the eventing meeting today (interview); is there a way you could see this becoming a "MUST", or do you think it always needs to be a MAY/SHOULD?

Reply:
I see it as always a MAY/SHOULD for 1.0